clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/ops/DeployCloudFoundryServerGroupAtomicOperation.java

```diff
@@ -120,45 +129,48 @@ private static CloudFoundryServerGroup createApplication(DeployCloudFoundryServe
+      description.getSpace(),
+      description.getApplicationAttributes().getBuildpacks(),
+      description.getApplicationAttributes().getEnv());
     getTask().updateStatus(PHASE, "Created Cloud Foundry application '" + description.getServerGroupName() + "'");
 
     return serverGroup;
   }
 
-  private String buildPackage(String serverGroupId, DeployCloudFoundryServerGroupDescription description) {
-    final CloudFoundryClient client = description.getClient();
-    getTask().updateStatus(PHASE, "Creating package for application '" + description.getServerGroupName() + "'");
-
-    final String packageId = client.getApplications().createPackage(serverGroupId);
-
+  @Nullable
+  private File downloadPackageArtifact(DeployCloudFoundryServerGroupDescription description) {
     File file = null;
     try {
-      InputStream artifactInputStream = description.getArtifactCredentials().download(description.getArtifact());
-      file = File.createTempFile(description.getArtifact().getReference(), null);
+      InputStream artifactInputStream = description.getArtifactCredentials().download(description.getApplicationArtifact());
+      file = File.createTempFile(UUID.randomUUID().toString(), null);
       FileOutputStream fileOutputStream = new FileOutputStream(file);
       IOUtils.copy(artifactInputStream, fileOutputStream);
       fileOutputStream.close();
-      client.getApplications().uploadPackageBits(packageId, file);
-
-      operationPoller.waitForOperation(
-        () -> client.getApplications().packageUploadComplete(packageId),
-        Function.identity(), null, getTask(), description.getServerGroupName(), PHASE);
-
-      getTask().updateStatus(PHASE, "Completed creating package for application '" + description.getServerGroupName() + "'");
     } catch (IOException e) {
+      file.delete();
       throw new UncheckedIOException(e);
-    } finally {
-      if (file != null) {
-        file.delete();
-      }
     }
+    return file;
+  }
+
+  private String buildPackage(String serverGroupId, DeployCloudFoundryServerGroupDescription description, File packageArtifact) {
+    CloudFoundryClient client = description.getClient();
+    getTask().updateStatus(PHASE, "Creating package for application '" + description.getServerGroupName() + "'");
+
+    String packageId = client.getApplications().createPackage(serverGroupId);
+    client.getApplications().uploadPackageBits(packageId, packageArtifact);
+
+    operationPoller.waitForOperation(
+      () -> client.getApplications().packageUploadComplete(packageId),
+      Function.identity(), null, getTask(), description.getServerGroupName(), PHASE);
+
+    getTask().updateStatus(PHASE, "Completed creating package for application '" + description.getServerGroupName() + "'");
 
     return packageId;
   }
 
   private void buildDroplet(String packageId, String serverGroupId, DeployCloudFoundryServerGroupDescription description) {
-    final CloudFoundryClient client = description.getClient();
```

Comment:
why remove final?

Reply:
why should final be on a local variable?